src/app/modules/webhook/webhook.message.ts

```diff
@@ -0,0 +1,174 @@
+import { Result } from 'neverthrow'
+
+import { createLoggerWithLabel } from '../../config/logger'
+
+import {
+  DUE_TIME_TOLERANCE_SECONDS,
+  QUEUE_MESSAGE_VERSION,
+} from './webhook.constants'
+import {
+  WebhookNoMoreRetriesError,
+  WebhookQueueMessageParsingError,
+} from './webhook.errors'
+import {
+  WebhookFailedQueueMessage,
+  webhookMessageSchema,
+  WebhookQueueMessageObject,
+  WebhookQueueMessagePrettified,
+} from './webhook.types'
+import { getNextAttempt, prettifyEpoch } from './webhook.utils'
+
+const logger = createLoggerWithLabel(module)
+
+/**
+ * Encapsulates a queue message for webhook retries.
+ */
+export class WebhookQueueMessage {
+  message: WebhookQueueMessageObject
+
+  constructor(message: WebhookQueueMessageObject) {
+    this.message = message
+  }
+
+  /**
+   * Converts a webhook queue message body into an encapsulated
+   * class instance.
+   * @param body Raw body of webhook queue message
+   * @returns ok(encapsulated message) if message can be parsed successfully
+   * @returns err if message fails to be parsed
+   */
+  static deserialise(
+    body: string,
+  ): Result<WebhookQueueMessage, WebhookQueueMessageParsingError> {
+    return Result.fromThrowable(
+      () => JSON.parse(body) as unknown,
+      (error) => {
+        logger.error({
+          message: 'Unable to parse webhook queue message body',
+          meta: {
+            action: 'deserialise',
+            body,
+          },
+          error,
+        })
+        return new WebhookQueueMessageParsingError(error)
+      },
+    )()
+      .andThen((parsed) =>
+        Result.fromThrowable(
+          () => webhookMessageSchema.parse(parsed),
+          (error) => {
+            logger.error({
+              message: 'Webhook queue message body has wrong shape',
+              meta: {
+                action: 'deserialise',
+                body,
+              },
+              error,
+            })
+            return new WebhookQueueMessageParsingError(error)
+          },
+        )(),
+      )
+      .map((validated) => new WebhookQueueMessage(validated))
+  }
+
+  /**
+   * Initialises a webhook queue message which has not been
+   * retried as yet. This function succeeds as long as
+   * the retry policy allows for at least one retry.
+   * @param submissionId
+   * @returns ok(encapsulated message) if retry policy exists
+   * @returns err if the retry policy does not allow any retries
+   */
+  static fromSubmissionId(
+    submissionId: string,
+  ): Result<WebhookQueueMessage, WebhookNoMoreRetriesError> {
+    return getNextAttempt([]).map(
+      (nextAttempt) =>
+        new WebhookQueueMessage({
+          submissionId,
+          previousAttempts: [],
+          nextAttempt,
+          _v: QUEUE_MESSAGE_VERSION,
+        }),
+    )
+  }
+
+  /**
+   * Serialises for enqueueing.
+   * @returns Serialised message
+   */
+  serialise(): string {
+    return JSON.stringify(this.message)
```

Comment:
do we want to use our custom stringify-safe instead?

Reply:
will do, but I actually don't think it's necessary, as we are in control of what we are stringifying here. `stringify-safe` is more for cases where the object we're stringifying comes from an external source.

Reply:
actually if it's okay with you, I'd rather just stick to `JSON.stringify`. `stringifySafe` can return `undefined` if the argument is `undefined`, and I'd rather not work around that unless there's a good reason to use `stringifySafe`.